R/data.table.R

```diff
@@ -1559,7 +1559,7 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
                                 jl__ = as.list(jsubl[[i_]])[-1L] # just keep the '.' from list(.)
                                 jn__ = if (is.null(names(jl__))) rep("", length(jl__)) else names(jl__)
                                 idx  = unlist(lapply(jl__, function(x) is.name(x) && x == ".I"))
-                                if (any(idx)) jn__[idx] = ifelse(jn__[idx] == "", "I", jn__[idx])
+                                if (any(idx)) jn__[idx & (jn__ == "")] = "I"
```

Comment:
Was initially worried that we're now evaluating `jn__ == ""` on the full vector, but a (rink-a-dink) benchmark suggests we're still doing much better:

```
library(microbenchmark)
set.seed(1023)
full <- as.character(sample(10))

idx_low <- sample(10, 4)
idx_high <- sample(10, 9)

full_low <- full; full_low[sample(idx_low, 1)] <- ""
full_high <- full; full_high[sample(idx_high, 6)] <- ""

idx_low <- 1:10 %in% idx_low
idx_high <- 1:10 %in% idx_high

times <- matrix(nrow = 1000, ncol = 6)

for (repetition in 1:1000){
  jn <- full_low
  beg <- get_nanotime()
  jn[idx_low & jn == ""] = "I"
  end <- get_nanotime()
  times[repetition, 1] <- end - beg
  
  jn <- full_low
  beg <- get_nanotime()
  jn[idx_low] = {o <- jn[idx_low]; o[o==""] <- "I"; o}
  end <- get_nanotime()
  times[repetition, 2] <- end - beg
  
  jn <- full_low
  beg <- get_nanotime()
  jn[idx_low] = ifelse(jn[idx_low] == "", "I", jn[idx_low])
  end <- get_nanotime()
  times[repetition, 3] <- end - beg
  
  jn <- full_high
  beg <- get_nanotime()
  jn[idx_high & jn == ""] = "I"
  end <- get_nanotime()
  times[repetition, 4] <- end - beg
  
  jn <- full_high
  beg <- get_nanotime()
  jn[idx_high] = {o <- jn[idx_high]; o[o==""] <- "I"; o}
  end <- get_nanotime()
  times[repetition, 5] <- end - beg
  
  jn <- full_high
  beg <- get_nanotime()
  jn[idx_high] = ifelse(jn[idx_high] == "", "I", jn[idx_high])
  end <- get_nanotime()
  times[repetition, 6] <- end - beg
}

apply(times, 2, median)
# [1]  2132.0  3117.5  9320.0  2160.0  3423.0 11056.0
```